j5/backends/hardware/sr/v4/motor_board.py

```diff
@@ -0,0 +1,190 @@
+            for _ in range(0, 2)
+        ]
+
+        self._serial = serial_class(
+            port=serial_port,
+            baudrate=1000000,
+            timeout=0.25,
+        )
+
+        # Check we have the correct firmware version.
+        version = self.get_firmware_version()
+        if version != "3":
+            raise CommunicationError(
+                f"Unexpected firmware version: {version}, expected: 3.",
+            )
+
+        # Brake both of the motors
+        for i, val in enumerate(self._state):
+            self.set_motor_state(i, val)
+
+    def __del__(self) -> None:
+        """Clean up device on destruction of object."""
+        # Brake both of the motors for safety
+        for i, val in enumerate(self._state):
+            self.set_motor_state(i, MotorSpecialState.BRAKE)
+
+        self._serial.flush()
+        self._serial.close()
+
+    @handle_serial_error
+    def send_command(self, command: int, data: Optional[int] = None) -> None:
+        """Send a serial command to the board."""
+        command_bytes = chr(command).encode('utf-8')
+        bytes_written = self._serial.write(command_bytes)
+        if len(command_bytes) != bytes_written:
+            raise CommunicationError(
+                "Mismatch in command bytes written to serial interface.",
+            )
+
+        if data is not None:
+            data_bytes = chr(data).encode()
+            bytes_written = self._serial.write(data_bytes)
+            # It is not possible to test the following if statement without refactor.
+            if len(data_bytes) != bytes_written:  # pragma: nocover
+                raise CommunicationError(
+                    "Mismatch in data bytes written to serial interface.",
+                )
+
+    @handle_serial_error
+    def read_serial_line(self) -> str:
+        """Read a line from the serial interface."""
+        bdata = self._serial.readline()
+
+        if len(bdata) == 0:
+            raise CommunicationError(
+                "Unable to communicate with motor board. ",
+                "Is it correctly powered?",
+            )
+
+        ldata = bdata.decode('utf-8')
+        return ldata.rstrip()
+
+    def get_firmware_version(self) -> Optional[str]:
+        """Get the firmware version of the board."""
+        self.send_command(CMD_VERSION)
+        firmware_data = self.read_serial_line()
+        model = firmware_data[:5]
+        if model != "MCV4B":
+            raise CommunicationError(
+                f"Unexpected model string: {model}, expected MCV4B.",
+            )
+        return firmware_data[6:]  # Strip model and return version
+
+    def get_motor_state(self, identifier: int) -> MotorState:
+        """Get the state of a motor."""
+        # We are unable to read the state from the motor board,
+        # so we'll get the last set value.
+        return self._state[identifier]
+
+    def set_motor_state(self, identifier: int, power: MotorState) -> None:
```

Comment:
Something to consider, instead of `set_motor_state`, what if this class actually exposed itself as a list of `self._state`, ie if you instantiatied this class and called it `foo`. `foo[0].motor_state = power`. and then the `motor_state` can be a property setter of the `MotorState` class.. or something along those lines at least?

Reply:
I personally like the concept as it is more pythonic, although I am unsure how it would work in this instance.

Remember, a `Backend` is never directly called by a competitor. It is an abstract interface that is called by the relevant `Component`. In this case, the `set_motor_state` function is called by the `Motor.state` setter.